File: solutions/javascript/blind-75/linked-list/reorderList.js
const getMiddleNode = (head) => {
    let slowNode = head;
    let fastNode = head;

    // The slow node will reach the middle of the list when the fast node reaches the end
    while (fastNode) {
        fastNode = fastNode.next;
        if (!fastNode) return slowNode;
        fastNode = fastNode.next;
        slowNode = slowNode.next;
    }

    return slowNode;
};

const reverseList = (head) => {
    let curNode = head;
    let prevNode = null;

    while (curNode) {
        // Keep the next node so that you don't lose it when overwriting it
        const nextNode = curNode.next;

        // Set the current node's next neighbor to previous node
        curNode.next = prevNode;

        // Move forward in the list
        prevNode = curNode;
        curNode = nextNode;
    }

    return prevNode;
};

const mergeLists = (listOneNode, listTwoNode) => {
    while (listTwoNode.next) {
        // We want to move the listNodeOne forward by one (i.e. listNodeOne = listNodeOne.next). However,
        // we also want to overwrite listNodeOne.next to be listNodeTwo. So, we store a reference to
        // listNodeOne.next BEFORE we overwrite it.
        let temporary = listOneNode.next;
        listOneNode.next = listTwoNode;
        listOneNode = temporary;

        // We do the same thing here but with listNodeTwo
        temporary = listTwoNode.next;
        listTwoNode.next = listOneNode;
        listTwoNode = temporary;
    }
};

const reorderList = (head) => {
    // We want to split the list into two halves, so we find the middle node
    const middleNode = getMiddleNode(head);

    // We reverse the second half of the list (i.e. n, n - 1, n - 2 ...)
    let reversedHalfNode = reverseList(middleNode);
    let firstHalfNode = head;

    // We then merge the first half of the list (i.e. 1, 2, 3, ...) and the second half of the list which was reversed (i.e. n, n - 1, n - 2 ...)
    mergeLists(firstHalfNode, reversedHalfNode);
};

File: solutions/python/general-algos/linked-lists/reorder_list.py


class Solution:
    def get_middle_node(self, head):
        slow_node = head
        fast_node = head

        # The slow node will reach the middle of the list when the fast node reaches the end
        while fast_node:
            fast_node = fast_node.next
            if not fast_node: return slow_node
            fast_node = fast_node.next
            slow_node = slow_node.next

        return slow_node

    def reverse_list(self, head):
        curNode = head
        prevNode = None

        while curNode:
            # Keep the next node so that you don't lose it when overwriting it
            nextNode = curNode.next

            # Set the current node's next neighbor to previous node
            curNode.next = prevNode

            # Move forward in the list
            prevNode = curNode
            curNode = nextNode

        return prevNode

    def merge_lists(self, list_one_node, list_two_node):
        while list_two_node.next:
            # We want to move the list_one_node forward by one (i.e. list_one_node = list_one_node.next). However,
            # we also want to overwrite list_one_node.next to be list_two_node. So, we store a reference to
            # list_one_node.next BEFORE we overwrite it.
            temporary = list_one_node.next
            list_one_node.next = list_two_node
            list_one_node = temporary

            # We do the same thing here but with list_two_node
            temporary = list_two_node.next
            list_two_node.next = list_one_node
            list_two_node = temporary

    def reorderList(self, head: Optional[ListNode]) -> None:
        # We want to split the list into two halves, so we find the middle node
        middleNode = self.get_middle_node(head)

        # We reverse the second half of the list (i.e. n, n - 1, n - 2 ...)
        reversedHalfNode = self.reverse_list(middleNode)
        firstHalfNode = head

        # We then merge the first half of the list (i.e. 1, 2, 3, ...) and the second half of the list which was reversed (i.e. n, n - 1, n - 2 ...)
        self.merge_lists(firstHalfNode, reversedHalfNode)

    

    